### PR TITLE
IGNITE-22735 Sql. Avoid double parsing of sql queries

### PR DESCRIPTION
**Pull request by korlov42**
https://issues.apache.org/jira/browse/IGNITE-22735

Results of e2e benchmark on my laptop as follow: 
```
Before:

Benchmark                          (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt    Score   Error  Units
InsertBenchmark.sqlInlinedInsert               1    false                25  avgt   20  249.793 ± 6.699  us/op
InsertBenchmark.sqlPreparedInsert              1    false                25  avgt   20   74.698 ± 5.611  us/op

After:

Benchmark                          (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt    Score   Error  Units
InsertBenchmark.sqlInlinedInsert               1    false                25  avgt   20  221.851 ± 6.699  us/op
InsertBenchmark.sqlPreparedInsert              1    false                25  avgt   20   74.618 ± 6.140  us/op
```

----------------

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImpl.java [hunk 1]
```diff
@@ -55,12 +56,23 @@ public ParsedResult parse(String query) {
...
+
+        @SuppressWarnings("UnnecessaryLocalVariable")
         ParsedResult result = new ParsedResultImpl(
                 queryType,
                 query,
                 normalizedQuery,
                 parsedStatement.dynamicParamsCount(),
-                () -> IgniteSqlParser.parse(query, StatementParseResult.MODE).statement()
+                () -> {
+                    SqlNode ast = holder.getAndSet(null);
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImpl.java:68]: hacker )

Comment (ygerzhedovich) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImpl.java:68]: Let's add a comment to explain what going on to future readers of the piece of code

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java [hunk 2]
```diff
@@ -322,4 +324,44 @@ private RelNode convertMerge(SqlMerge call) {
                 relBuilder.build(), LogicalTableModify.Operation.MERGE,
                 targetColumnNameList, null, false);
     }
+
+    // =========================================================
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:328]: I suggest create a ticket with a patch for Calcite and add mention the ticket here

**Review by ygerzhedovich — APPROVED** [review]

**Review by zstan — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java (hunk 2 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:328]: done

https://issues.apache.org/jira/browse/IGNITE-22755

**Pull request merged**